superset-frontend/src/dashboard/components/nativeFilters/FilterCard/useFilterScope.ts

```diff
@@ -0,0 +1,144 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Filter, t } from '@superset-ui/core';
+import { useSelector } from 'react-redux';
+import {
+  ChartsState,
+  Layout,
+  LayoutItem,
+  RootState,
+} from 'src/dashboard/types';
+import { DASHBOARD_ROOT_ID } from 'src/dashboard/util/constants';
+import { CHART_TYPE } from 'src/dashboard/util/componentTypes';
+import { useMemo } from 'react';
+
+const useCharts = () => {
+  const charts = useSelector<RootState, ChartsState>(state => state.charts);
+  return useMemo(() => Object.values(charts), [charts]);
+};
+
+export const useFilterScope = (filter: Filter) => {
+  const layout = useSelector<RootState, Layout>(
+    state => state.dashboardLayout.present,
+  );
+  const charts = useCharts();
+
+  return useMemo(() => {
+    let topLevelTabs: string[] | undefined;
+    if (layout[DASHBOARD_ROOT_ID].children[0].startsWith('TABS-')) {
+      topLevelTabs = layout[layout[DASHBOARD_ROOT_ID].children[0]].children;
+    }
+
+    // no charts in scope
+    if (filter.scope.rootPath.length === 0) {
+      return undefined;
+    }
+
+    // all charts in scope
+    // no charts excluded and no top level tabs
+    // OR no charts excluded and every top level tab is in rootPath
+    if (
+      filter.scope.excluded.length === 0 &&
+      (filter.scope.rootPath[0] === DASHBOARD_ROOT_ID ||
+        (topLevelTabs &&
+          topLevelTabs.every(topLevelTab =>
+            filter.scope.rootPath.includes(topLevelTab),
+          )))
+    ) {
+      return [t('All charts')];
+    }
+
+    // no charts excluded and not every top level tab in scope
+    // returns "TAB1, TAB2"
+    if (filter.scope.excluded.length === 0 && topLevelTabs) {
+      return filter.scope.rootPath.map(
+        tabId => layout[tabId].meta.text || layout[tabId].meta.defaultText,
+      );
+    }
+
+    const layoutCharts = Object.values(layout).filter(
+      layoutElement => layoutElement.type === CHART_TYPE,
+    );
+
+    // no top level tabs, charts excluded
+    // returns "CHART1, CHART2"
+    if (filter.scope.rootPath[0] === DASHBOARD_ROOT_ID) {
+      return charts
+        .filter(chart => !filter.scope.excluded.includes(chart.id))
+        .map(chart => {
+          const layoutElement = layoutCharts.find(
+            layoutChart => layoutChart.meta.chartId === chart.id,
+          );
+          return (
+            layoutElement?.meta.sliceNameOverride ||
+            layoutElement?.meta.sliceName ||
+            layoutElement?.id
+          );
+        })
+        .filter(Boolean);
+    }
+
+    // top level tabs, charts excluded
+    // returns "TAB1, TAB2, CHART1"
+    if (topLevelTabs) {
+      const topLevelTabsInFullScope = [...filter.scope.rootPath];
```

Comment:
Maybe add a comment like `We start assuming that all charts are in scope for all tabs in the root path.`